### PR TITLE
New case:RHEL-187695 - Snapshot on read-only drive

### PR DESCRIPTION
**Pull request by aliang123**
Signed-off-by: Aihua Liang <aliang@redhat.com>
id:1979805

### CODE REVIEW

**Comment by aliang123** [comment]
(1/1) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.5.0.x86_64.io-github-autotest-qemu.blockdev_snapshot_readonly.q35: PASS (65.75 s)

**Comment by aliang123** [comment]
@zhencliu, @ngu-niny, please help to review, thanks.

File: qemu/tests/blockdev_snapshot_readonly.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        try:
+            mount_point = utils_disk.configure_empty_linux_disk(
+                session, disk_id, disk_size)[0]
+        except ShellCmdError as e:
+            disk_tag = r"/dev/%s" % disk_id
+            error_msg = self.params['error_msg'] % disk_tag
+            if error_msg not in str(e):
+                self.test.fail("Unexpected disk format error: %s" % str(e))
+            self.disks_info[self.base_tag] = [disk_tag, '/mnt']
+        finally:
```
Comment (zhencliu) [inline qemu/tests/blockdev_snapshot_readonly.py:26]: we may add 'else' before finally,  to catch the issue that the read-only disk can be modified unexpectedly.

File: qemu/tests/blockdev_snapshot_readonly.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+def run(test, params, env):
+    """
+    Backup VM disk test when VM reboot
+
+    1) start VM with data disk and read-only
+    2) check block info before snapshot
+    3) do snapshot to target disk
+    4) check block info after snapshot
+    5) check data disk in guest
```
Comment (zhencliu) [inline qemu/tests/blockdev_snapshot_readonly.py:64]: Can we update the steps to keep consistent with the steps we do in script?
e.g.  We rebase snapshot onto the data disk after VM down, then we clone the VM and try to mount the data disk in rw way, which should fail

File: qemu/tests/cfg/blockdev_snapshot_readonly.cfg [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+    snapshot_tag = sn1
+    image_format_sn1 = qcow2
+    image_name_sn1 = images/sn1
+    image_size_sn1 = ${image_size_data1}
+    image_readonly_sn1 = yes
+    device = "drive_data1"
+    base_tag = "data1"
+    rebase_mode = unsafe
+    error_msg = "Warning: Unable to open %s read-write (Read-only file system)"
```
Comment (zhencliu) [inline qemu/tests/cfg/blockdev_snapshot_readonly.cfg:23]: Just fyi.
I used to hit the issue that different msgs showed when configuring a read-only disk.

File: qemu/tests/blockdev_snapshot_readonly.py (hunk 1 above)
Comment (aliang123) [reply qemu/tests/blockdev_snapshot_readonly.py:30]: Have updated, please help to check again, zhenchao.

File: qemu/tests/blockdev_snapshot_readonly.py (hunk 2 above)
Comment (aliang123) [reply qemu/tests/blockdev_snapshot_readonly.py]: Have upated, thanks zhenchao.

File: qemu/tests/cfg/blockdev_snapshot_readonly.cfg (hunk 3 above)
Comment (aliang123) [reply qemu/tests/cfg/blockdev_snapshot_readonly.cfg]: Have updated it, please help to review again. zhenchao

File: qemu/tests/cfg/blockdev_snapshot_readonly.cfg [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    snapshot_tag = sn1
+    image_format_sn1 = qcow2
+    image_name_sn1 = images/sn1
+    image_size_sn1 = ${image_size_data1}
+    image_readonly_sn1 = yes
+    device = "drive_data1"
+    base_tag = "data1"
+    rebase_mode = unsafe
+    error_msg = "Warning: Unable to open %s read-write"
+    error_msg + = "|because it is opened read-only"
```
Comment (zhencliu) [inline qemu/tests/cfg/blockdev_snapshot_readonly.cfg:24]: + =  --->  +=, no space between + and =

**Comment by zhencliu** [comment]
LGTM

Just remove the space between + and =

File: qemu/tests/cfg/blockdev_snapshot_readonly.cfg (hunk 4 above)
Comment (aliang123) [reply qemu/tests/cfg/blockdev_snapshot_readonly.cfg]: Updated it, thanks zhenchao.

File: qemu/tests/cfg/blockdev_snapshot_readonly.cfg [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    snapshot_tag = sn1
+    image_format_sn1 = qcow2
+    image_name_sn1 = images/sn1
+    image_size_sn1 = ${image_size_data1}
+    image_readonly_sn1 = yes
+    device = "drive_data1"
+    base_tag = "data1"
+    rebase_mode = unsafe
+    error_msg = "Warning: Unable to open %s read-write"
+    error_msg += "|because it is opened read-only"
```
Comment (ngu-niny) [inline qemu/tests/cfg/blockdev_snapshot_readonly.cfg:24]: @aliang123 For virtio-blk disk, the error prompt is as following:

[root@dhcp16-215-194 ~]# parted -s "/dev/vdb" mklabel msdos
Error: Operation not permitted during write on /dev/vdb
[root@dhcp16-215-194 ~]#

**Comment by ngu-niny** [comment]
PASS 1-Host_RHEL.m8.u5.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_readonly

Acked-by: Nini Gu <ngu@redhat.com>

**Pull request merged**